### PR TITLE
Fix error in tensor_index.hpp for 3+ qubit operations

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

There was an error in the simulator tensor index function when applying 3+ qubit operations. Such operations are only applied when measurement sampling is called for a subset of qubits. This pull request fixes the indexing function so it works correctly for n-qubit operations.

This PR also includes an updated makefile for the simulator that can be used to enable OpenMP parallelization with Apple builds. It now preferences building with Apple Xcode builds if the LLVM OpenMP library is installed using Homebrew (`brew install libomp`), failing that it will build with GCC if that can be found, and failing that builds with XCode without openmp.

### CODE REVIEW

**Comment by chriseclectic** [comment]
I found another bug related to n-qubit gates. This one doesn't effect any current use cases of the simulator, but it needed to be fixed. A `default:` guard was missing for the last case of a several `switch` statements that would potentially cause the gate to be applied twice.

File: src/qasm-simulator-cpp/src/utilities/tensor_index.hpp [hunk 1]
```diff
@@ -100,29 +93,32 @@ TensorIndex::TensorIndex() {
...
 uint_t TensorIndex::index0_static(const std::array<uint_t, N> &qubits_sorted,
                                  const uint_t k) const {
-  uint_t lowbits = 0, mask = 0;
+  uint_t lowbits, retval = k;
   for (size_t j = 0; j < N; j++) {
-    mask ^= masks[qubits_sorted[j] - j];
-    lowbits |= (k & mask) << j;
+    lowbits = retval & masks[qubits_sorted[j]];
+    retval >>= qubits_sorted[j];
+    retval <<= qubits_sorted[j] + 1;
```
Comment (atilag) [inline src/qasm-simulator-cpp/src/utilities/tensor_index.hpp]: I can't see the difference between these two shifts:
``` 
    retval >>= qubits_sorted[j];
    retval <<= qubits_sorted[j] + 1;
```
and just:
```
retval <<= 1;
```
Am I wrong?

File: src/qasm-simulator-cpp/src/utilities/tensor_index.hpp [hunk 2]
```diff
@@ -160,23 +158,24 @@ TensorIndex::indexes_static(const std::array<uint_t, 2> &qs,
...
+                                   const size_t N,
                                    const uint_t k) const {
-  uint_t lowbits = 0, mask = 0;
+  uint_t lowbits, retval = k;
   for (size_t j = 0; j < N; j++) {
-    mask ^= masks[qubits_sorted[j] - j];
-    lowbits |= (k & mask) << j;
+    lowbits = retval & masks[qubits_sorted[j]];
+    retval >>= qubits_sorted[j];
+    retval <<= qubits_sorted[j] + 1;
```
Comment (atilag) [inline src/qasm-simulator-cpp/src/utilities/tensor_index.hpp]: Same as above here?

File: src/qasm-simulator-cpp/src/utilities/tensor_index.hpp (hunk 1 above)
Comment (chriseclectic) [reply src/qasm-simulator-cpp/src/utilities/tensor_index.hpp]: It's necessary. We are taking an arbitrary bit-string and inserting a 0 into a specified location by:
* shifting down to clear lower bits (`retval >>= qubits_sorted[j]`)
* shifting up to insert a bit leaving it and all lowbits in zero (`retval <<= qubits_sorted[j] + 1`). 
* Then the lowbits are reset to initial value (`retval |= lowbits`)
Just shifting would insert the zero bit at the start shifting all bits over 1.

Comment (atilag) [reply src/qasm-simulator-cpp/src/utilities/tensor_index.hpp]: wow, ok!

**Pull request merged**